### PR TITLE
fix metric sort error 

### PR DESCRIPTION
**Pull request by linuxgcc**
#1442

### CODE REVIEW

**Comment by ArthurSens** [comment]
Although I agree that sometimes humans debug exposed metrics by manually looking the /metrics endpoint, I believe this endpoint it focused on metrics being scraped by Prometheus.

If we introduce a regex sort, I wonder how much are we sacrificing in performance for this change. Do you mind doing some benchmarking and posting the results here?

**Comment by linuxgcc** [comment]
@ArthurSens 
Sorry, I don't have time recently, benchmark has been completed.

1 . touch such file  prometheus/internal/metric_sort_test.go
2. with out reg sort benchmark
> [root@trace client_golang]# go test -bench="BenchmarkSort" -benchtime=100x -benchmem  ./prometheus/internal                                                                                                         
> goos: linux
> goarch: amd64
> pkg: github.com/prometheus/client_golang/prometheus/internal
> cpu: Intel Xeon Processor (Cascadelake)
> BenchmarkSort-16             100              2020 ns/op             265 B/op          8 allocs/op
> PASS
> ok      github.com/prometheus/client_golang/prometheus/internal 0.012s

4. with reg sort benchmark

> [root@trace client_golang]# go test -bench="BenchmarkSort" -benchtime=100x -benchmem  ./prometheus/internal                                                                                                         
> goos: linux
> goarch: amd64
> pkg: github.com/prometheus/client_golang/prometheus/internal
> cpu: Intel Xeon Processor (Cascadelake)
> BenchmarkSort-16             100             88625 ns/op           52093 B/op        626 allocs/op
> PASS
> ok      github.com/prometheus/client_golang/prometheus/internal 0.022s
> 
> [root@trace client_golang]# 

It can be seen from the above comparison that the comparison of 16-core CPU data before and after using regular expressions takes no more than 0.1ms

**Comment by ArthurSens** [comment]
We can also see a huge increase in memory allocation though 😱

File: prometheus/internal/metric.go [hunk 1]
```diff
@@ -46,6 +48,21 @@ func (s MetricSorter) Swap(i, j int) {
 	s[i], s[j] = s[j], s[i]
 }
 
+func (s MetricSorter) compare(si, sj string) bool {
+	// When the number of devices exceeds two digits, a sorting error occurs.
+	// dealing with sorting problems, especially CPU sorting errors
+	re := regexp.MustCompile(`\d+$`)
```
Comment (bwplotka) [inline prometheus/internal/metric.go:54]: Hm I think we can't use regex in such a hot path, we might need to be more smarter here if we ever do this ( let's talk on the issue) - or do it optionally.

**Review by bwplotka — CHANGES_REQUESTED** [review]
Nice, thanks for this!

I added more points to the discussion https://github.com/prometheus/client_golang/issues/1442#issuecomment-2046695361

@ArthurSens is right, as you said it's "only" 0.1ms but if ops is one comparison.. we do this potentially 10 000 times per scrape making it a second. Number of allocs is one thing (a lot more) but the bump from 200 B to 50Kb is not minor (per comparison!). Maybe I am misinterpreting what "ops" is in your benchmark, maybe it's per full sorting (but for how many series?) In future please paste the benchmark code too, so we can verify.

But performance is one thing, the most important part is:
*  if readability hit of NOT having this is big enough to make this change
* if numerical sorting would NOT actually surprise (thus impact readability) of ppl who got used to it

**Comment by bwplotka** [comment]
Sounds like we have to reject this, sorry! This is an intended behaviour.

**Pull request closed without merging**